File: multidim_with_der/levenberg_marquadt.py
from typing import Callable, Dict, List
import numpy as np
import numpy.typing as npt 
# Import my functionalities:
import sys
sys.path.append("../useful_python_scripts/")  
from useful_plots import plot_level_curves_and_points
from derivative_funcitons import gradient_of_fun_in_point, hessian_matrix
 
"""

"""

def forward_substitution(L, b):
    """
    Solves the linear system Lx = b for x using forward substitution,
    where L is a lower triangular matrix.
    """
    n = L.shape[0]
    x = np.zeros_like(b, dtype=float)
    for i in range(n):
        x[i] = (b[i] - np.dot(L[i, :i], x[:i])) / L[i, i]
    return x

def backward_substitution(U, b):
    """
    Solves the linear system Ux = b for x using backward substitution,
    where U is an upper triangular matrix.
    """
    n = U.shape[0]
    x = np.zeros_like(b, dtype=float)
    for i in range(n-1, -1, -1):
        x[i] = (b[i] - np.dot(U[i, i+1:], x[i+1:])) / U[i, i]
    return x

def levenberg_marquadt(
    epsilon: float,
    initial_point: npt.ArrayLike,
    delta: float,
    obj_function: Callable[[npt.ArrayLike], float],  
    max_iter: int = 500, 
) -> Dict:
    """
    Levenberg-Marquardt optimization algorithm for minimizing a function.

    Parameters:
        epsilon (float): Convergence criterion. Terminate when the magnitude of the new movement
                         is less than epsilon.
        initial_point (array_like): Initial guess for the optimal solution.
        delta (float): Initial value for the parameter delta. (expected to be short)
        obj_function (callable): The objective function to minimize.
        max_iter (int, optional): Maximum number of iterations. Defaults to 500.

    Returns:
        dict: A dictionary containing:
              - 'value': The optimized solution.
              - 'converged': Boolean indicating if the algorithm converged.
              - 'iterations': Number of iterations performed.
              - 'trayectory': List of all points visited during optimization.
    """

    # Set up
    iterations: int = 0
    last_point: npt.ArrayLike = initial_point.copy()
    all_points_trayectory: List[npt.ArrayLike] = [initial_point]

    while iterations < max_iter:  
        # Calculate the gradient in the point
        gradient: npt.ArrayLike = gradient_of_fun_in_point(obj_function, last_point)
        # Calculate the hessian matrix of the fun in the point
        hessian: npt.ArrayLike = hessian_matrix(obj_function, last_point)

        # Calculate B^-1 = delta*I + H(xk)
        B_m1 = delta*np.eye(len(hessian)) + hessian
        # Cholesky factorization
        L = np.linalg.cholesky(B_m1) 
        # Solve LL' (X_{k+l} - X_k) = -vf(x_k) for X_{k+l}
        Y: npt.ArrayLike = forward_substitution(L, -1*gradient)         #vec
        new_point: npt.ArrayLike = backward_substitution(L.T, Y) + last_point

        # Add the point to the trayectory
        all_points_trayectory.append(new_point.copy())  

        # Compute f(x_{k+1}), f(x_k) and ratio R_k
        f_k = obj_function(last_point)
        f_kl = obj_function(new_point)
        q_k = f_k + np.dot(gradient, new_point - last_point) + 0.5 * np.dot(np.dot((new_point - last_point).T, hessian), new_point - last_point)
        q_kl = f_kl
        R_k = (f_k - f_kl) / (q_k - q_kl)

        # Update ep_k+l based on R_k
        if R_k < 0.25:
            delta *= 4
        elif R_k > 0.75:
            delta /= 2
 
        # Calculate the next movement of the point
        new_mov: npt.ArrayLike = new_point - last_point
        # End the algorithm if the magnitude of the new movement is lower that epsilon
        magnitude: float = np.linalg.norm(new_mov)
        if magnitude < epsilon:
            break 
        # update for next levenberg_marquadt iteration
        last_point = new_point.copy()
        # cont the iteration
        iterations += 1

    # Check if max iterations achieved
    is_good_sol: bool = iterations < max_iter
    return {
        "value": last_point,
        "converged": is_good_sol,
        "iterations": iterations,
        "trayectory": all_points_trayectory,
    }


def function_R2(x: npt.ArrayLike) -> float:
    """
    Example function to test.
    For f: R2 -> R
    """
    x1 = x[0]
    x2 = x[1]
    return (x1 - 2) ** 4 + (x1 - (2 * x2)) ** 2


def function_R2_b(x: npt.ArrayLike) -> float:
    """
    Example function to test.
    For f: R2 -> R
    """
    x1 = x[0]
    x2 = x[1]
    return (x1 - 2) ** 2 + (x2 - 4) ** 2

def main() -> None:
    # Define a termination scalar, to define the threshold of how much is close enough for the solution
    epsilon: float = 0.001
    # Define an arbitrary point
    # initial_point: npt.ArrayLike = np.array([-2.1, 7.0])
    initial_point: npt.ArrayLike = np.array([0.0, 3.0])
    # The function to minize:
    fun = function_R2 
    # Call the function for cyclic coordinate algorithm using golden section method.
    solution: Dict = levenberg_marquadt(epsilon, initial_point, 0.5, fun)
    # Just print solution
    print(f"The solution is: {solution["value"]}")
    # Print the trayectory length
    print(f"Points in trayectory: {len(solution["trayectory"])}")

    if solution["converged"]:
        # Make the level curves plot
        plot_level_curves_and_points(solution["trayectory"],fun)
    
if __name__ == "__main__":
    main() 

File: miniproyect/matlab/levenberg_marquadt/levenberg_marquadt.m
function solution = levenberg_marquadt(epsilon, initial_point, obj_function, max_iter)
    % Levenberg-Marquardt optimization algorithm for minimizing a function.
    % 
    % Parameters:
    % epsilon (float): Convergence criterion. Terminate when the magnitude of the new movement
    %                  is less than epsilon.
    % initial_point (array_like): Initial guess for the optimal solution.
    % obj_function (callable): The objective function to minimize.
    % max_iter (int, optional): Maximum number of iterations. Defaults to 100.
    % 
    % Returns:
    % struct: A struct containing:
    %   - value: The optimized solution.
    %   - converged: Boolean indicating if the algorithm converged.
    %   - iterations: Number of iterations performed.
    %   - trayectory: List of all points visited during optimization.

    if nargin < 4
        max_iter = 150;
    end

    % Set up
    initial_delta = 0.8;
    delta = initial_delta;
    iterations = 0;
    last_point = initial_point;
    all_points_trayectory = {initial_point};

    while iterations < max_iter
        % Increment the iteration counter
        iterations = iterations + 1;
        % Calculate the gradient at the point
        fprintf("\nIter %d: Calculating gradient",iterations) 
        gradient = gradient_of_fun_in_point(obj_function, last_point,epsilon)';
        magnitude = norm(gradient); 
        % End the algorithm if the magnitude of the new movement is less than epsilon
        fprintf('! |gradient| = %f\n', magnitude);
        if magnitude < epsilon
            break;
        end
        % Calculate the Hessian matrix of the function at the point
        fprintf("\nIter %d: Calculating hessian matrix",iterations)
        hessian = hessian_matrix(obj_function, last_point, epsilon);
        
        while true
            try
                % Calculate B^-1 = delta*I + H(xk)
                B_m1 = delta * eye(length(hessian)) + hessian;
                % Cholesky factorization
                L = chol(B_m1, 'lower');
                break;
            catch
                delta = delta * 4;
            end
        end
        
        % Solve LL' (X_{k+l} - X_k) = -vf(x_k) for X_{k+l} 
        Y = L \ (-gradient);  % vector 
        step = ((L') \ Y);
        step = epsilon*step./norm(step);
        new_point = last_point + step;
        
        % Add the point to the trajectory
        all_points_trayectory{end+1} = new_point;
        
        % Compute f(x_{k+1}), f(x_k) and ratio R_k
        f_k = obj_function(last_point);
        f_kl = obj_function(new_point);
        %fprintf("SNDR(new_point) = %f",f_kl)
        q_k = f_k + dot(gradient, new_point - last_point) + 0.5 * (new_point - last_point)' * hessian * (new_point - last_point);
        q_kl = f_kl;
        R_k = (f_k - f_kl) / (q_k - q_kl);
        
        % Update delta based on R_k
        if R_k < 0.25
            delta = delta * 4;
        elseif R_k > 0.75
            delta = delta / 2;
        end

        % Update for next Levenberg-Marquardt iteration
        last_point = new_point;
    end
    
    % Check if max iterations achieved
    is_good_sol = iterations < max_iter;
    solution = struct('value', last_point, 'converged', is_good_sol, 'iterations', iterations, 'trayectory', {all_points_trayectory}, 'initial_delta', initial_delta);
end

function hess = hessian_matrix(fun, point, epsilon)
    % Approximate numerically the Hessian matrix of a function evaluated at a given point.
    
    if nargin < 3
        epsilon = 1e-5;
    end
    
    % Get the number of dimensions
    num_dim = length(point);
    
    % Initialize Hessian matrix
    hess = zeros(num_dim, num_dim);
    
    % Calculate each element of the Hessian matrix using central difference
    for i = 1:num_dim
        for j = 1:num_dim
            % Perturb the point along the i-th and j-th axes
            perturbed_point1 = point;
            perturbed_point1(i) = perturbed_point1(i) + epsilon;
            perturbed_point1(j) = perturbed_point1(j) + epsilon;
            
            perturbed_point2 = point;
            perturbed_point2(i) = perturbed_point2(i) + epsilon;
            perturbed_point2(j) = perturbed_point2(j) - epsilon;
            
            perturbed_point3 = point;
            perturbed_point3(i) = perturbed_point3(i) - epsilon;
            perturbed_point3(j) = perturbed_point3(j) + epsilon;
            
            perturbed_point4 = point;
            perturbed_point4(i) = perturbed_point4(i) - epsilon;
            perturbed_point4(j) = perturbed_point4(j) - epsilon;
            
            % Calculate central difference
            hess(i, j) = (fun(perturbed_point1) - fun(perturbed_point2) - fun(perturbed_point3) + fun(perturbed_point4)) / (4 * epsilon^2);
        end
    end
end

function gradient = gradient_of_fun_in_point(fun, point, epsilon)
    % Approximate the value of the gradient of a function at a given point using central difference method.
    
    if nargin < 3
        epsilon = 1e-6;
    end
    
    % Define the dimensions
    num_dim = length(point);
    
    % Define a unitary vector in each dimension
    directions = eye(num_dim);
    
    % Calculate the gradient by differentiating in each direction
    gradient = arrayfun(@(i) derivate_fun_in_point_and_dir(fun, point, directions(:, i), epsilon), 1:num_dim);
    % for i = 1:num_dim
    %     gradient(i) = derivate_fun_in_point_and_dir(fun, point, directions(:, i), epsilon);
    % end
end

function d = derivate_fun_in_point_and_dir(fun, point, direction, epsilon)
    % Approximate the value of the derivative of a function at a given point in a given direction using central difference method.
    
    if nargin < 4
        epsilon = 1e-6;
    end
    
    % Check the direction has unitary magnitude
    if norm(direction) ~= 1.0
        direction = (1 / norm(direction)) * direction;
    end
    
    point_left = point - (epsilon * direction);
    point_right = point + (epsilon * direction);
    
    d = (fun(point_right) - fun(point_left)) / (2 * epsilon);
end 